From 2519feb99f60ee3d313dd02548c71b9153723b46 Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Fri, 25 Apr 2025 15:32:05 +0800
Subject: [PATCH 103/107] mtk: mac80211: fix radio idx missing after replace
 ctx

The radio idx of the new ctx would become -1 after channel switch.
This leads to using ieee80211_link_use_reserved_reassign for the next
channel switch while the chanctx list is actually full, especially for
mt7992 and mt7990, so ieee80211_vif_use_reserved_switch should be used instead.

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 net/mac80211/chan.c | 21 ++++++++++++++++++---
 1 file changed, 18 insertions(+), 3 deletions(-)

diff --git a/net/mac80211/chan.c b/net/mac80211/chan.c
index 48e53aa5..8aa286f7 100644
--- a/net/mac80211/chan.c
+++ b/net/mac80211/chan.c
@@ -1153,11 +1153,13 @@ static struct ieee80211_chanctx *
 ieee80211_replace_chanctx(struct ieee80211_local *local,
 			  const struct ieee80211_chan_req *chanreq,
 			  enum ieee80211_chanctx_mode mode,
-			  struct ieee80211_chanctx *curr_ctx)
+			  struct ieee80211_chanctx *curr_ctx,
+			  u32 radio_mask)
 {
 	struct ieee80211_chanctx *new_ctx, *ctx;
 	struct wiphy *wiphy = local->hw.wiphy;
 	const struct wiphy_radio *radio;
+	int i, radio_idx = -1;
 
 	if (!curr_ctx || (curr_ctx->replace_state ==
 			  IEEE80211_CHANCTX_WILL_BE_REPLACED) ||
@@ -1207,7 +1209,19 @@ ieee80211_replace_chanctx(struct ieee80211_local *local,
 	    !list_empty(&curr_ctx->reserved_links))
 		return ERR_PTR(-EBUSY);
 
-	new_ctx = ieee80211_alloc_chanctx(local, chanreq, mode, -1);
+	for (i = 0; i < wiphy->n_radio; i++) {
+		if (!(radio_mask & BIT(i)))
+			continue;
+
+		radio = &wiphy->radio[i];
+		if (!cfg80211_radio_chandef_valid(radio, &chanreq->oper))
+			continue;
+
+		radio_idx = i;
+		break;
+	}
+
+	new_ctx = ieee80211_alloc_chanctx(local, chanreq, mode, radio_idx);
 	if (!new_ctx)
 		return ERR_PTR(-ENOMEM);
 
@@ -1279,7 +1293,8 @@ int ieee80211_link_reserve_chanctx(struct ieee80211_link_data *link,
 							false, radio_idx);
 		else
 			new_ctx = ieee80211_replace_chanctx(local, chanreq,
-							    mode, curr_ctx);
+							    mode, curr_ctx,
+							    sdata->wdev.radio_mask);
 		if (IS_ERR(new_ctx))
 			return PTR_ERR(new_ctx);
 	}
-- 
2.45.2

